File: modules/download-wake-lock/android/src/main/java/expo/modules/downloadwakelock/DownloadWakeLockModule.kt
package expo.modules.downloadwakelock

import android.content.Context
import android.os.PowerManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * One PARTIAL_WAKE_LOCK (CPU on, screen free to turn off) held while model
 * downloads run. Not reference-counted: acquire while held is a no-op and a
 * single release frees it, so repeated calls can never stack locks. JS
 * (src/services/downloadWakeLock.ts) decides when downloads start and end.
 */
class DownloadWakeLockModule : Module() {
  private var wakeLock: PowerManager.WakeLock? = null

  override fun definition() = ModuleDefinition {
    Name("DownloadWakeLock")

    Function("acquire") {
      acquire()
    }

    Function("release") {
      release()
    }

    // JS reload or app teardown: nothing on the JS side will release it anymore.
    OnDestroy {
      release()
    }
  }

  private fun acquire(): Boolean {
    val lock = wakeLock ?: run {
      val context = appContext.reactContext ?: return false
      val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
      pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, TAG).apply { setReferenceCounted(false) }
    }.also { wakeLock = it }
    // Timeout is a backstop only; JS releases it when the download settles.
    if (!lock.isHeld) lock.acquire(MAX_HOLD_MS)
    return true
  }

  private fun release() {
    wakeLock?.let { if (it.isHeld) it.release() }
  }

  companion object {
    const val TAG = "BOAR:ModelDownload"
    const val MAX_HOLD_MS = 6L * 60 * 60 * 1000
  }
}

File: modules/download-wake-lock/ios/DownloadWakeLockModule.swift
import ExpoModulesCore
import UIKit

/**
 * iOS counterpart of DownloadWakeLockModule.kt. iOS has no CPU wake lock, so
 * "keep the download going" is two things here:
 *
 * - isIdleTimerDisabled: the screen does not auto-lock while a download runs,
 *   so the app stays in the foreground with the user watching progress.
 * - beginBackgroundTask: if the user leaves the app anyway, iOS grants a short
 *   grace period (typically ~30s) before suspending it.
 *
 * The transfer itself survives suspension because expo-file-system's
 * createDownloadResumable uses a background URLSession on iOS by default
 * (sessionType BACKGROUND); this module does not need to own the session.
 * Same contract as Android: not reference-counted, acquire while held is a
 * no-op, one release frees it.
 */
public class DownloadWakeLockModule: Module {
  private var held = false
  private var backgroundTask: UIBackgroundTaskIdentifier = .invalid

  public func definition() -> ModuleDefinition {
    Name("DownloadWakeLock")

    Function("acquire") { () -> Bool in
      self.onMain { self.acquire() }
      return true
    }

    Function("release") {
      self.onMain { self.release() }
    }

    // JS reload or app teardown: nothing on the JS side will release it anymore.
    OnDestroy {
      self.onMain { self.release() }
    }
  }

  private func acquire() {
    guard !held else { return }
    held = true
    UIApplication.shared.isIdleTimerDisabled = true
    backgroundTask = UIApplication.shared.beginBackgroundTask(withName: "BOAR:ModelDownload") { [weak self] in
      // Grace period over: iOS will suspend us. End the task so it doesn't
      // kill the app; the background URLSession keeps transferring.
      self?.endBackgroundTask()
    }
  }

  private func release() {
    guard held else { return }
    held = false
    UIApplication.shared.isIdleTimerDisabled = false
    endBackgroundTask()
  }

  private func endBackgroundTask() {
    guard backgroundTask != .invalid else { return }
    UIApplication.shared.endBackgroundTask(backgroundTask)
    backgroundTask = .invalid
  }

  private func onMain(_ block: @escaping () -> Void) {
    if Thread.isMainThread { block() } else { DispatchQueue.main.async(execute: block) }
  }
}
